### PR TITLE
Patient model

### LINKED ISSUE #60: Add patient model
- patient model should extend person
- change person to abstract
- write tests

### PR DESCRIPTION
**Pull request by owen-yap**
- TODO: Add tests
- current model sufficient for basic commands, need to add patient list and doctor list to relevant models once we start working on patient assignment.
- Closes #60

### CODE REVIEW

File: src/main/java/seedu/address/model/person/patient/Height.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+import static seedu.address.commons.util.AppUtil.checkArgument;
+
+/**
+ * Represents a Patient's height in the address book.
+ * Guarantees: immutable; is valid as declared in {@link #isValidHeight(String)}
+ */
+public class Height {
+
+    public static final String MESSAGE_CONSTRAINTS =
+            "Height should be a positive decimal point number representing the patient's height in metres "
```
Comment (praveenkrishna0512) [inline src/main/java/seedu/address/model/person/patient/Height.java:13]: By the way, should we account for the diagnosis beginning with white spaces too? Or will our logic trim away leading and trailing whitespaces?

File: src/main/java/seedu/address/model/person/patient/Patient.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    private final Diagnosis diagnosis;
+    private final Status status;
+    private final Remark remark;
+    /**
+     * Every field must be present and not null.
+     */
+    public Patient(Name name, Phone phone, Email email, Height height, Weight weight,
+                   Diagnosis diagnosis, Status status, Remark remark, Set<Tag> tags) {
+        super(name, phone, email, tags);
+        requireAllNonNull(name, phone, email, height, weight, diagnosis, status, remark, tags);
```
Comment (praveenkrishna0512) [inline src/main/java/seedu/address/model/person/patient/Patient.java:30]: Do we need to check `requireAllNonNull` here for `name, phone, email` and `tags` if `super(name, phone, email, tags)` will do that for us?

File: src/main/java/seedu/address/model/person/patient/Status.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+ * Represents a Patient's hospitalization status in the address book.
+ * Guarantees: immutable; is valid as declared in {@link #isValidStatus(String)}
+ */
+public class Status {
+
+    public static final String MESSAGE_CONSTRAINTS =
+            "Status should be one of the following: Inpatient, Outpatient, Observation, Emergency Department,"
+                    + " Intensive Care Unit, Transitional Care";
+
+    public static final String VALIDATION_REGEX = "(?i)^(Inpatient|Outpatient|Observation|Emergency Department"
```
Comment (praveenkrishna0512) [inline src/main/java/seedu/address/model/person/patient/Status.java:15]: Regex looks good!

File: src/main/java/seedu/address/model/person/patient/Weight.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+/**
+ * Represents a Patient's weight in the address book.
+ * Guarantees: immutable; is valid as declared in {@link #isValidWeight(String)}
+ */
+public class Weight {
+
+    public static final String MESSAGE_CONSTRAINTS =
+            "Weight should be in kilograms (kg), expressed as a positive decimal number.";
+
+    public static final String VALIDATION_REGEX = "^[0-9]+(\\.[0-9]{1,2})?$";
```
Comment (praveenkrishna0512) [inline src/main/java/seedu/address/model/person/patient/Weight.java:15]: This regex seems to fail for decimal numbers with floating point on https://regex101.com/. It works with integers though.

Comment (praveenkrishna0512) [reply src/main/java/seedu/address/model/person/patient/Weight.java:15]: Should we also check for leading and trailing whitespaces?

**Review by praveenkrishna0512 — COMMENTED** [review]
Logic looks good to me for the most part! Just require a few RegEx changes here and there :) Thanks!

File: src/main/java/seedu/address/model/person/patient/Status.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    private final String status;
+
+    /**
+     * Constructs a {@code Status}.
+     *
+     * @param status A valid hospitalization status.
+     */
+    public Status(String status) {
+        requireNonNull(status);
+        this.status = status.substring(0, 1).toUpperCase() + status.substring(1).toLowerCase();
```
Comment (Bobfree546) [inline src/main/java/seedu/address/model/person/patient/Status.java:27]: Should the capitalization only be for the first letter of the string or every word?

File: src/main/java/seedu/address/model/person/patient/Height.java (hunk 1 above)
Comment (Bobfree546) [reply src/main/java/seedu/address/model/person/patient/Height.java:13]: Duplicate comment

File: src/main/java/seedu/address/model/person/patient/Diagnosis.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+public class Diagnosis {
+
+    public static final String MESSAGE_CONSTRAINTS =
+            "Diagnosis should only contain alphanumeric characters and spaces, and it should not be blank";
+
+    /*
+     * The first character of the address must not be a whitespace,
+     * otherwise " " (a blank string) becomes a valid input.
+     */
+    public static final String VALIDATION_REGEX = "[\\p{Alnum}][\\p{Alnum} ]*";
```
[the remark this answers is not in the document]
Comment (owen-yap) [reply src/main/java/seedu/address/model/person/patient/Diagnosis.java:19]: This regex is the same as name, should work. I can write tests for this on Sunday

File: src/main/java/seedu/address/model/person/patient/Remark.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Represents any remarks for a person in the address book.
+ * Guarantees: immutable; is always valid
+ */
+public class Remark {
+
+    public static final String MESSAGE_CONSTRAINTS = "Add any remarks for the patient.";
```
[the remark this answers is not in the document]
Comment (owen-yap) [reply src/main/java/seedu/address/model/person/patient/Remark.java:11]: I think we need to discuss this feature, probably want a way to add / remove remarks as well. Lets KIV for now

File: src/main/java/seedu/address/model/person/patient/Diagnosis.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+import static seedu.address.commons.util.AppUtil.checkArgument;
+
+/**
+ * Represents a Patient's diagnosis in the address book.
+ * Guarantees: immutable; is valid as declared in {@link #isValidDiagnosis(String)}
+ */
+public class Diagnosis {
+
+    public static final String MESSAGE_CONSTRAINTS =
+            "Diagnosis should only contain alphanumeric characters and spaces, and it should not be blank";
```
[the remark this answers is not in the document]
Comment (owen-yap) [reply src/main/java/seedu/address/model/person/patient/Diagnosis.java:13]: whitespaces not allowed, we can change this later

**Review by Bobfree546 — APPROVED** [review]
LGTM

**Pull request merged**